### PR TITLE
EC/CUDA: enable memops for executor

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Enable wait kernel and memops when executor is used only for stream sync

### CODE REVIEW

**Comment by Sergei-Lebedev** [comment]
addresses #672

**Comment by edgargabriel** [comment]
If I understand the pr correctly, this pr changes the ec interfaces and breaks ROCm support. Something to keep track of.

**Comment by Sergei-Lebedev** [comment]
> If I understand the pr correctly, this pr changes the ec interfaces and breaks ROCm support. Something to keep track of.

Yes, you are right. I can remove those functions from ROCm component, they are not used anyway, wdyt?

**Comment by edgargabriel** [comment]
@Sergei-Lebedev ok, sounds good, ping me when you have that ready and I can run your branch through our internal CI

File: src/components/ec/cuda/ec_cuda.h [hunk 1]
```diff
@@ -115,6 +97,7 @@ typedef struct ucc_ec_cuda_executor_task_ops {
 typedef struct ucc_ec_cuda_executor {
     ucc_ee_executor_t                super;
     ucc_ec_cuda_executor_mode_t      mode;
+    uint64_t                         requested_ops;
     ucc_ec_cuda_executor_task_ops_t  ops;
     ucc_spinlock_t                   tasks_lock;
     ucc_ec_cuda_executor_state_t     state;
```
Comment (samnordmann) [inline src/components/ec/cuda/ec_cuda.h:103]: not related to this PR, but shouldn't `state` need to be allocated with `cudaHostAlloc`?

**Review by samnordmann — APPROVED** [review]

File: src/components/ec/cuda/ec_cuda_executor.c [hunk 2]
```diff
@@ -25,6 +30,13 @@ ucc_status_t ucc_cuda_executor_init(const ucc_ee_executor_params_t *params,
         return UCC_ERR_NO_MEMORY;
     }
 
+    if (params->mask & UCC_EE_EXECUTOR_PARAM_FIELD_TASK_TYPES) {
+        eee->requested_ops = params->task_types;
+    } else {
+        /* if no task types provided assume all tasks types required */
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/ec/cuda/ec_cuda_executor.c:36]: in next PR I'm going to add different task_types for different collectives e.g. reduce task and copy task for reduce_scatter in TL/CUDA

File: src/components/ec/cuda/ec_cuda_executor.c [hunk 3]
```diff
@@ -25,6 +30,13 @@ ucc_status_t ucc_cuda_executor_init(const ucc_ee_executor_params_t *params,
...
+        eee->requested_ops = 1;
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/ec/cuda/ec_cuda_executor.c:37]: right, it's currently just placeholder for next PR, I can remove it until then

File: src/components/ec/cuda/ec_cuda.h (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/components/ec/cuda/ec_cuda.h:103]: it's allocated with cudaHostAlloc see ucc_ec_cuda_ee_executor_mpool_ops

**Comment by Sergei-Lebedev** [comment]
@edgargabriel i removed unused function from EC/ROCM, please test

**Comment by edgargabriel** [comment]
@Sergei-Lebedev thank you, I can confirm that things still work and look. Thanks!

**Review by manjugv — APPROVED** [review]

**Review by bureddy — APPROVED** [review]

**Comment by yangrudan** [comment]
> ## What
> Enable wait kernel and memops when executor is used only for stream sync

Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.

My command is as below:

**UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall  -M cuda -m 1024 --triggered 1**

Looking forward to your reply, thank you.

**Comment by yangrudan** [comment]
> ## What
> Enable wait kernel and memops when executor is used only for stream sync

Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.

My command is as below:

**UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall  -M cuda -m 1024 --triggered 1**

Looking forward to your reply, thank you~

**Comment by Sergei-Lebedev** [comment]
> > ## What
> > Enable wait kernel and memops when executor is used only for stream sync
> 
> Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> 
> My command is as below:
> 
> **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> 
> Looking forward to your reply, thank you.

I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.

**Comment by yangrudan** [comment]
> > > ## 什么
> > > 当 executor 仅用于流同步时，启用 wait kernel 和 memops
> > 
> > 
> > 您能告诉我在什么情况下会使用 **wait kernel 或 memops** 吗？我尝试运行测试工具，但似乎无法访问 ec_cuda_executor_persistent_wait.c 文件中的函数。
> > 我的命令如下：
> > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > 期待您的回复，谢谢。
> 
> 我不知道什么是 zccl_test_mpi，但是当用户需要流式有序集合作时，会使用 wait_kernel 或 memops，并且这个集合作实现不需要执行任何 CUDA作（例如 memcpy、reduce）。在这种情况下，如果设备和驱动程序支持 memops，则 UCC 将使用 memops 并等待内核。



> > > ## What
> > > Enable wait kernel and memops when executor is used only for stream sync
> > 
> > 
> > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > My command is as below:
> > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > Looking forward to your reply, thank you.
> 
> I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.

Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.

I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.

Thank you very much.

**Comment by ikryukov** [comment]
> > > > ## 什么
> > > > 当 executor 仅用于流同步时，启用 wait kernel 和 memops
> > > 
> > > 
> > > 您能告诉我在什么情况下会使用 **wait kernel 或 memops** 吗？我尝试运行测试工具，但似乎无法访问 ec_cuda_executor_persistent_wait.c 文件中的函数。
> > > 我的命令如下：
> > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > 期待您的回复，谢谢。
> > 
> > 
> > 我不知道什么是 zccl_test_mpi，但是当用户需要流式有序集合作时，会使用 wait_kernel 或 memops，并且这个集合作实现不需要执行任何 CUDA作（例如 memcpy、reduce）。在这种情况下，如果设备和驱动程序支持 memops，则 UCC 将使用 memops 并等待内核。
> 
> > > > ## What
> > > > Enable wait kernel and memops when executor is used only for stream sync
> > > 
> > > 
> > > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > > My command is as below:
> > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > Looking forward to your reply, thank you.
> > 
> > 
> > I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.
> 
> Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.
> 
> I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.
> 
> Thank you very much.

Hello, you could find launch command to test collectives with memops in #1093 
`mpirun --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=<user>/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCX_TLS=rc,cuda -x UCX_RNDV_THRESH=0 -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_perftest -m cuda -c bcast -F -d uint8 -b 1 -e 8k -T`

**Comment by moranhhuishou1995** [comment]
> > > > > ## What
> > > > > Enable wait kernel and memops when executor is used only for stream sync
> > > > 
> > > > 
> > > > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > > > My command is as below:
> > > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > > Looking forward to your reply, thank you.
> > > 
> > > 
> > > I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.
> > 
> > 
> > Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.
> > I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.
> > Thank you very much.
> 
> Hello, you could find launch command to test collectives with memops in #1093 `mpirun --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=<user>/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCX_TLS=rc,cuda -x UCX_RNDV_THRESH=0 -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_perftest -m cuda -c bcast -F -d uint8 -b 1 -e 8k -T`


Hello, when I use this command to test collectives, the function ucc_cuda_executor_persistent_wait_start() will be executed, but when I change UCC_TLS=ucp to UCC_TLS=cuda, the ucc_cuda_executor_persistent_start() will be executed, does tl/cuda supports memops?
image

**Comment by moranhhuishou1995** [comment]
> > > > > > ## What
> > > > > > Enable wait kernel and memops when executor is used only for stream sync
> > > > > 
> > > > > 
> > > > > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > > > > My command is as below:
> > > > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > > > Looking forward to your reply, thank you.
> > > > 
> > > > 
> > > > I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.
> > > 
> > > 
> > > Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.
> > > I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.
> > > Thank you very much.
> > 
> > 
> > Hello, you could find launch command to test collectives with memops in #1093 `mpirun --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=<user>/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCX_TLS=rc,cuda -x UCX_RNDV_THRESH=0 -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_perftest -m cuda -c bcast -F -d uint8 -b 1 -e 8k -T`
> 
> Hello, when I use this command to test collectives, the function ucc_cuda_executor_persistent_wait_start() will be executed, but when I change UCC_TLS=ucp to UCC_TLS=cuda, the ucc_cuda_executor_persistent_start() will be executed, does tl/cuda supports memops? image

By the way, my collective type is **alltoall.**

**Comment by Sergei-Lebedev** [comment]
> > > > > > ## What
> > > > > > Enable wait kernel and memops when executor is used only for stream sync
> > > > > 
> > > > > 
> > > > > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > > > > My command is as below:
> > > > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > > > Looking forward to your reply, thank you.
> > > > 
> > > > 
> > > > I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.
> > > 
> > > 
> > > Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.
> > > I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.
> > > Thank you very much.
> > 
> > 
> > Hello, you could find launch command to test collectives with memops in #1093 `mpirun --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=<user>/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCX_TLS=rc,cuda -x UCX_RNDV_THRESH=0 -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_perftest -m cuda -c bcast -F -d uint8 -b 1 -e 8k -T`
> 
> Hello, when I use this command to test collectives, the function ucc_cuda_executor_persistent_wait_start() will be executed, but when I change UCC_TLS=ucp to UCC_TLS=cuda, the ucc_cuda_executor_persistent_start() will be executed, does tl/cuda supports memops? image

TL CUDA needs persistent executor to move data between ranks using kernel copy. TL UCP alltoall doesn't need persistent executor since in some scenarios it can use gpu direct rdma, so memops is enough to guarantee stream synchronization in TL UCP.

**Comment by moranhhuishou1995** [comment]
> Thanks for your reply

Thanks for your reply. Does the current code support the use of memops in TL CUDA? What should I do if i want to use memops in TL CUDA？

**Comment by moranhhuishou1995** [comment]
> > > > > > > ## What
> > > > > > > Enable wait kernel and memops when executor is used only for stream sync
> > > > > > 
> > > > > > 
> > > > > > Could you tell me in what scenarios **wait kernel or memops** would be used? I tried running the testing tool, but it seems I cannot reach the functions in the ec_cuda_executor_persistent_wait.c file.
> > > > > > My command is as below:
> > > > > > **UCC_COLL_TRACE=info mpirun --allow-run-as-root -n 2 ./zccl_test_mpi -c alltoall -M cuda -m 1024 --triggered 1**
> > > > > > Looking forward to your reply, thank you.
> > > > > 
> > > > > 
> > > > > I don't know what is zccl_test_mpi, but wait_kernel or memops used when user needs stream ordered collective operation and this collective operation implementation doesn't need to do any CUDA operations (e.g. memcpy, reduce). In this case if memops is supported by device and driver then UCC will use memops and wait kernel otherwise.
> > > > 
> > > > 
> > > > Thanks for your reply. Is a spelling mistake. My test tool is ucc_test_mpi, and it's full path is ucc/build/bin/ucc_test_mpi.
> > > > I understand what you mean, can you give me a test or use case? Let me experiment with MemOps. As a novice, I am not sufficiently familiar with UCC.
> > > > Thank you very much.
> > > 
> > > 
> > > Hello, you could find launch command to test collectives with memops in #1093 `mpirun --mca coll ^hcoll --mca coll_ucc_enable 0 -x LD_LIBRARY_PATH=<user>/ucc/install/lib:$LD_LIBRARY_PATH -x UCC_CLS=basic -x UCC_TLS=ucp -x UCX_TLS=rc,cuda -x UCX_RNDV_THRESH=0 -x UCC_LOG_LEVEL=info -x UCC_CONFIG_FILE= -np 2 ./install/bin/ucc_perftest -m cuda -c bcast -F -d uint8 -b 1 -e 8k -T`
> > 
> > 
> > Hello, when I use this command to test collectives, the function ucc_cuda_executor_persistent_wait_start() will be executed, but when I change UCC_TLS=ucp to UCC_TLS=cuda, the ucc_cuda_executor_persistent_start() will be executed, does tl/cuda supports memops? image
> 
> TL CUDA needs persistent executor to move data between ranks using kernel copy. TL UCP alltoall doesn't need persistent executor since in some scenarios it can use gpu direct rdma, so memops is enough to guarantee stream synchronization in TL UCP.

Thanks for your reply. Does the current code support the use of memops in TL CUDA? What should I do if i want to use memops in TL CUDA？

**Pull request merged**